google-cloud-tools-plugin/src/com/google/gct/idea/debugger/CloudBreakpointHandler.java

```diff
@@ -440,16 +440,16 @@ public void unregisterBreakpoint(
       boolean temporary) {
     // If the state was set to disabled as a result of a server update,
     // then we do not need to update the server side.
-    if (xIdeBreakpoint.getProperties().isDisabledByServer()) {
-      return;
-    }
-    String breakpointId = xIdeBreakpoint.getUserData(CLOUD_ID);
-    if (!Strings.isNullOrEmpty(breakpointId)) {
-      assert breakpointId != null;
-      myProcess.getStateController().deleteBreakpointAsync(breakpointId);
-    }
-    else {
-      LOG.warn("could not delete breakpoint because it was not added through the cloud handler.");
+    if (!xIdeBreakpoint.getProperties().isDisabledByServer()) {
+      String breakpointId = xIdeBreakpoint.getUserData(CLOUD_ID);
+      if (!Strings.isNullOrEmpty(breakpointId)) {
+        myProcess.getStateController().deleteBreakpointAsync(breakpointId);
+      } else {
+        LOG.warn("could not delete breakpoint because it was not added through the cloud handler.");
+      }
     }
+    // reset this flag: either it has been disabled by the server or the client has deleted it, in
+    // both cases we need to add it again, if it is re-enabled
+    xIdeBreakpoint.getProperties().setAddedOnServer(false);
```

Comment:
What happens here when you add a breakpoint, stop the session, resume the session, stop the session, hit the breakpoint offline and resume the session again? Does the original bug resurface?